tinygrad/tensor.py

```diff
@@ -1961,8 +1962,23 @@ def interpolate(self, size:Tuple[int, ...], mode:str="linear", align_corners:boo
     ```
     """
     assert isinstance(size, (tuple,list)) and all_int(size) and 0 < len(size) <= self.ndim, f"invalid {size=}"
-    assert mode == "linear", "only supports linear interpolate"
+    assert not (mode == "nearest" and align_corners), 'align_corners option can only be set with the mode="linear"'
     x, expand = self, list(self.shape)
+    if mode in {"nearest", "nearest-exact"}:
+      arrs = (Tensor.arange(s, dtype=dtypes.int32, device=self.device) for s in size)
+      scales = (self.shape[i] / size[i] for i in range(-len(size), 0))
+      # NOTE: for nearest-exact...
+      # https://github.com/pytorch/pytorch/commit/6adbe044e39c8e8db158d91e151aa6dead6e9aa4#diff-bf10b066c2c62c193fa16ea880093284cbb3e53a3406b702759cb31041e6ae91
+      #   // index_f32 = (output_index + 0.5) * scale - 0.5
+      # // input_index = round(index_f32)
+      # // Same as Pillow and Scikit-Image/Scipy ndi.zoom
+      # const int64_t src_index =
+      # std::min(static_cast<int64_t>(floorf((dst_index + 0.5) * scale)), input_size - 1);
+      idxs = [(arr*scale).floor().cast(dtypes.int) if mode == "nearest" else (((arr+0.5)*scale)).floor().cast(dtypes.int)
```

Comment:
for some reason the psuedo-code they gave for nearest exact is 
```
      // index_f32 = (output_index + 0.5) * scale - 0.5
      // input_index = round(index_f32)
```

but the actual cpp implementation is `((output_index + 0.5) * scale).floor()` ...
I don't get it... 

but tests pass with the actual implementation
results also match `PIL.NEAREST` I think.
Gonna spam a few more tests and then try to clean up.